File: algorithms/binarysearch.py
#!/usr/bin/env python
#--coding:utf-8--
import unittest

'''
loop search
'''
def search(seq, v, low, high):
	while low <= high:
		mid = (low + high) // 2
		if v == seq[mid]:
			return mid
		elif v > seq[mid]:
			low = mid + 1
		else:
			high = mid - 1
	return None

'''
recursive search
'''
def search2(seq, v, low , high):
	if low > high:
		return None
	mid = (low + high) // 2
	if v == seq[mid]:
		return mid
	elif v > seq[mid]:
		return search2(seq, v, mid + 1, high)
	else:
		return search2(seq, v, low, mid - 1)

class TestSearch(unittest.TestCase):
	def __init__(self, *args, **kwargs):
		super(TestSearch, self).__init__(*args, **kwargs)
		self.seq = [1,1.4,2,3,4,5,6,7,8,9,10]

	def test_search_should_return_2(self):
		v = 2
		result = search(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 2)

	def test_search_should_return_10(self):
		v = 10
		result = search(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 10)

	def test_search_should_return_0(self):
		v = 1
		result = search(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 0)

	def test_search2_should_return_2(self):
		v = 2
		result = search2(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 2)

	def test_search2_should_return_10(self):
		v = 10
		result = search2(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 10)

	def test_search2_should_return_0(self):
		v = 1
		result = search2(self.seq, v, 0, len(self.seq))
		self.assertEqual(result, 0)

if __name__ == '__main__':
	suite = unittest.TestLoader().loadTestsFromTestCase(TestSearch)
	unittest.TextTestRunner(verbosity=2).run(suite)

File: algorithms/binarysearch.java
import java.util.Arrays;

/**
 * Created by caden on 2017/5/4.
 */
public class SearchMain {
    public static void main(String[] args) {
        int[] seq = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        int result = search(seq, 5, 0, seq.length);
        System.out.println(result);
        result = search2(seq, 5, 0, seq.length);
        System.out.println(result);
        result = Arrays.binarySearch(seq,0,seq.length,5);
        System.out.println(result);
    }


    public static int search(int[] seq, int v, int low, int high) {
        while (low <= high) {
            int mid = (low + high) / 2;
            if (v == seq[mid]) {
                return mid;
            } else if (v > seq[mid]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return Integer.MIN_VALUE;
    }

    public static int search2(int[] seq, int v, int low, int high) {
        if (low > high) {
            return Integer.MIN_VALUE;
        }
        int mid = (low + high) / 2;
        if (v == seq[mid]) {
            return mid;
        } else if (v > seq[mid]) {
            return search2(seq, v, mid + 1, high);
        } else {
            return search2(seq, v, low, mid - 1);
        }
    }
}
